crates/libs/windows/src/core/heap.rs

```diff
@@ -19,11 +19,16 @@ pub unsafe fn heap_free(ptr: RawPtr) {
     }
 }
 
-pub fn heap_string<T: Copy + Default + Sized>(value: &[T]) -> *const T {
-    let buffer = heap_alloc((value.len() + 1) * std::mem::size_of::<T>()).expect("Could not allocate string") as *mut T;
-    let slice = unsafe { std::slice::from_raw_parts_mut(buffer, value.len() + 1) };
-    let (string, terminator) = slice.split_at_mut(value.len());
-    string.copy_from_slice(value);
-    terminator[0] = T::default();
-    buffer
+/// Copy a slice of `T` into a freshly allocated buffer with an additional default `T` at the end.
+///
+/// Returns a pointer to the beginning of the buffer
+pub fn heap_string<T: Copy + Default + Sized>(slice: &[T]) -> *const T {
+    unsafe {
+        let buffer = heap_alloc((slice.len() + 1) * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
+        for offset in 0..slice.len() {
+            std::ptr::write(buffer.add(offset), slice[offset]);
+        }
```

Comment:
This feels like it would be slower than just calling `memcpy` but hopefully the optimizer can figure that out (LLVM usually does).

Reply:
Instead of this loop you probably want `std::ptr::copy_nonoverlapping`, which will lower to a correct memcpy usage.

```rust
std::ptr::copy_nonoverlapping(slice.as_ptr(), buffer, slice.len());
```